src/main/kotlin/com/shopify/promises/Promise.kt

```diff
@@ -257,7 +257,15 @@ class Promise<T, E> private constructor(state: State<T, E>) {
   }
 
   private fun complete(result: Promise.Result<T, E>) {
-    val oldState = state.get() as? State.InProgress ?: return
+    val oldState = state.get()?.let {
+      when (it) {
+        is State.Idle -> throw IllegalStateException("Cannot mark an idle promise as completed.")
```

Comment:
Technically this is impossible as the only place where we able to resolve / reject promise is when we in `InProgress` state. So I think we can simplify it to just:
```
when(it) {
    is State.Complete -> throw IllegalStateException("Cannot complete a promise twice.")
    is State.InProgress -> it
    else -> null
}
```

Reply:
Yeah that's what I thought. But that means that if we do somehow end up in this state, something really went wrong badly. I felt it was safer to fail early to add some diagnostic info, rather than continuing quietly